README.md
![build](https://github.com/georapbox/shuffle-letters/workflows/build/badge.svg)
[![npm version](https://img.shields.io/npm/v/shuffle-letters.svg)](https://www.npmjs.com/package/shuffle-letters)
[![Coverage Status](https://coveralls.io/repos/github/georapbox/shuffle-letters/badge.svg?branch=main)](https://coveralls.io/github/georapbox/shuffle-letters?branch=main)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://georapbox.mit-license.org/@2021)

# shuffle-letters.js

A JavaScript library to shuffle the text content of a DOM element with an animated effect.

> NOTE: This library is a port to vanilla JavaScript of the [Shuffle Letters Effect](https://tutorialzine.com/2011/09/shuffle-letters-effect-jquery) jQuery plugin.

## Demo

Check [here](https://georapbox.github.io/shuffle-letters/) for a live demo.

## Install

```sh
$ npm install shuffle-letters --save
```

## Usage

The library is exported in UMD, CommonJS, and ESM formats. You can import it the following ways:

### Using ESM import statement

```js
import shuffleLetters from 'shuffle-letters';
```

### Using CommonJS require statement

```js
const shuffleLetters = require('shuffle-letters');

// If you use a bundler like Webpack, you may need to import it the following way 
// as it might try to use the ESM module instead of the CommonJS.
const shuffleLetters = require('shuffle-letters').default;
```

### Old school browser global
```html
<script src="https://unpkg.com/shuffle-letters"></script>
```

## API

```js
shuffleLetters(element, [config])
```

### Params

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| element | <code>HTMLElement</code> |  | The HTML element whose `textContent` we want to apply the shuffle effect. |
| [config] | <code>object</code> | <code>{}</code> | Configurable object to override the default options. |
| [config.text] | <code>string</code> | <code>&quot;&#x27;&#x27;&quot;</code> | Alternative text that can be used instead of the `textContent` of the `element`. |
| [config.iterations] | <code>number</code> | <code>8</code> | The number of times the characters will be shuffled before the animation ends. |
| [config.fps] | <code>number</code> | <code>30</code> | The amount of frames per second that the animation runs. |
| [config.onComplete] | <code>function</code> | <code>() &#x3D;&gt; void</code> | A callback function that is called when the animation of the effect is complete. |

**Returns**: <code>function</code> - Returns a function that when called, it clears the `timeoutID` which identifies the timer created by the call to `setTimeout()` that is used internally for the shuffle effect. Useful for cleanup purposes, for example when you want to remove the element before the animation is completed.

## Usage examples

### Example 1 - Default options
#### HTML
```html
<h1>This is a funcy title</h1>
```

#### JavaScript
```js
  shuffleLetters(document.querySelector('h1'));
```

### Example 2 - Custom options

#### HTML
```html
<h1>This is a funcy title</h1>
```

#### JavaScript
```js
shuffleLetters(document.querySelector('h1'), {
  text: 'Yet another funcy title',
  iterations: 12,
  fps: 60,
  onComplete: el => {
    console.log('Effect was completed.');
    console.log(el);
  }
});
```

### Example 3 - Shuffle many elements

#### HTML
```html
<ul>
  <li>List item 1</li>
  <li>List item 2</li>
  <li>List item 3</li>
  <li>List item 4</li>
</ul>
```

#### JavaScript
```js
const list = document.querySelector('ul');
const listItems = list.querySelectorAll('li');

Array.prototype.forEach.call(listItems, (element, index) => {
  shuffleLetters(element, {
    onComplete: elm => {
      if (index === listItems.length - 1) {
        console.log('All effects were completed');
      }
    }
  });
});
```

### Example 4 - Cleanup

#### HTML
```html
<h1>This is a funcy title</h1>
```

#### JavaScript
```js
  const clearShuffleLetters = shuffleLetters(document.querySelector('h1'), {
    iterations: 200
  });
  
  // Cleanup after 1 second
  setTimeout(() => clearShuffleLetters(), 1000);
```

## License

[The MIT License (MIT)](https://georapbox.mit-license.org/@2021)


CHANGELOG.md
# CHANGELOG

# v1.1.0 - 2023-08-21

- Separate digits from letters in the shuffle process. This makes the effect more predictable and consistent.

# v1.0.2 - 2021-10-07

- Minor update in documentation

# v1.0.1 - 2021-10-06

- Correct the instructions for requiring the CommonJS export in documentation.

# v1.0.0 - 2021-05-31

- Initial release


test/shuffleLetters.spec.js
import { JSDOM } from 'jsdom';
import { shuffleLetters } from '../src/shuffleLetters';
import { isElement } from '../src/isElement';

jest.mock('../src/isElement', () => ({
  isElement: jest.fn()
}));

describe('shuffle-letters.js', () => {
  const defaultText = 'Lorem ipusm dolor sit amet 123456 !@#$%^&*()';
  const { document } = new JSDOM(`<!DOCTYPE html><h1>${defaultText}</h1>`).window;
  const el = document.querySelector('h1');

  const options = {
    iterations: 1,
    fps: 60
  };

  beforeEach(() => {
    isElement.mockImplementation(() => true);
  });

  it('should throw error if first argument is not HTMLElement', () => {
    isElement.mockImplementation(() => false);

    expect(() => {
      return shuffleLetters(el);
    }).toThrow(new TypeError('Expected element to be a valid HTML element.'));
  });

  it('should call the onComplete callback', done => {
    const fn = jest.fn();

    shuffleLetters(el, {
      ...options,
      onComplete: el => {
        done();
        fn(el);
        expect(fn).toHaveBeenCalledWith(el);
      }
    });
  });

  it('the final text content of the element should be the same as in the beginning', done => {
    shuffleLetters(el, {
      ...options,
      onComplete: el => {
        done();
        expect(el.textContent).toBe(defaultText);
      }
    });
  });

  it('should override the element\'s default text content', done => {
    shuffleLetters(el, {
      ...options,
      text: 'Custom text',
      onComplete: el => {
        done();
        expect(el.textContent).toBe('Custom text');
      }
    });
  });

  it('should cleanup when called', () => {
    const clearShuffleLetters = shuffleLetters(el);

    expect(typeof clearShuffleLetters).toBe('function');

    clearShuffleLetters();
  });
});


rollup.config.js
import babel from '@rollup/plugin-babel';
import { terser } from 'rollup-plugin-terser';
import pkg from './package.json';

const LIBRARY_FILE_NAME = 'shuffle-letters';
const LIBRARY_NAME = 'shuffleLetters';
const EXTERNAL = []; // Indicate which modules should be treated as external
const GLOBALS = {}; // https://rollupjs.org/guide/en/#outputglobals

const banner = `/*!
 * ${pkg.name}
 * ${pkg.description}
 *
 * @version v${pkg.version}
 * @author ${pkg.author}
 * @homepage ${pkg.homepage}
 * @repository ${pkg.repository.url}
 * @license ${pkg.license}
 */`;

const makeConfig = (env = 'development') => {
  let bundleSuffix = '';

  if (env === 'production') {
    bundleSuffix = 'min.';
  }

  const config = {
    input: 'src/index.js',
    external: EXTERNAL,
    output: [
      {
        banner,
        name: LIBRARY_NAME,
        file: `dist/${LIBRARY_FILE_NAME}.umd.${bundleSuffix}js`, // UMD
        format: 'umd',
        exports: 'auto',
        globals: GLOBALS
      },
      {
        banner,
        file: `dist/${LIBRARY_FILE_NAME}.cjs.${bundleSuffix}js`, // CommonJS
        format: 'cjs',
        // We use `default` here as we are only exporting one thing using `export default`.
        // https://rollupjs.org/guide/en/#outputexports
        exports: 'default',
        globals: GLOBALS
      },
      {
        banner,
        file: `dist/${LIBRARY_FILE_NAME}.esm.${bundleSuffix}js`, // ESM
        format: 'es',
        exports: 'auto',
        globals: GLOBALS
      }
    ],
    plugins: [
      babel({
        babelHelpers: 'bundled',
        exclude: ['node_modules/**']
      })
    ]
  };

  if (env === 'production') {
    config.plugins.push(terser({
      output: {
        comments: /^!/
      }
    }));
  }

  return config;
};

export default commandLineArgs => {
  const configs = [
    makeConfig()
  ];

  // Production
  if (commandLineArgs.environment === 'BUILD:production') {
    configs.push(makeConfig('production'));
  }

  return configs;
};


src/getRandomCharacterForType.js
/**
 * Get a pseudo-random character from an available pool depending on the type of the character.
 *
 * @param {string} charType The type of the character. Available values are "lowerCase", "upperCase" and "symbol".
 * @returns {string} Returns the pseudo-random character.
 */
export function getRandomCharacterForType(charType) {
  let charsPool = '';

  switch (charType) {
    case 'lowerCase':
      charsPool = 'abcdefghijklmnopqrstuvwxyz';
      break;
    case 'upperCase':
      charsPool = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
      break;
    case 'digit':
      charsPool = '0123456789';
      break;
    case 'symbol':
      charsPool = ',.?/\\(^)![]{}*&^%$#\'"';
      break;
    default:
      // Handle a default case if needed
      break;
  }

  const charsPoolArray = charsPool.split('');

  return charsPoolArray[Math.floor(Math.random() * charsPoolArray.length)];
}


src/shuffleLetters.js
import { getRandomCharacterForType } from './getRandomCharacterForType';
import { isElement } from './isElement';

/**
 * Shuffle the text content of a DOM element
 *
 * @param {HTMLElement} element The HTML element whose `textContent` we want to apply the shuffle effect.
 * @param {object} [config={}] Configurable object to override the default options.
 * @param {string} [config.text=''] Alternative text that can be used instead of the `textContent` of the `element`.
 * @param {number} [config.iterations=8] The number of times the characters will be shuffled before the animation ends.
 * @param {number} [config.fps=30] The amount of frames per second that the animation runs.
 * @param {function} [config.onComplete=() => void] A callback function that is called when the animation of the effect is complete.
 * @returns {function} Returns a function that when called, it clears the `timeoutID` which identifies the timer created by the call to `setTimeout()` that is used internally for the shuffle effect.
 */
export function shuffleLetters(element, config = {}) {
  /**
   * Default options, if not provided by the user.
   * @type {object}
   */
  const defaults = {
    text: '',
    iterations: 8,
    fps: 30,
    onComplete: () => void 0
  };

  const options = {
    ...defaults,
    ...config
  };

  if (!isElement(element)) {
    throw new TypeError('Expected element to be a valid HTML element.');
  }

  /**
   * Holds all characters of the string to process.
   * @type {string[]}
   */
  const charsArray = options.text && typeof options.text === 'string'
    ? options.text.split('')
    : element.textContent.split('');

  /**
   * Holds the type for each character.
   * @type {string[]}
   */
  const charsTypes = [];

  /**
   * Holds the positions of non-space characters.
   * @type {number[]}
   */
  const charsPositions = [];

  for (let i = 0; i < charsArray.length; i += 1) {
    const char = charsArray[i];

    if (/\s/.test(char)) {
      charsTypes[i] = 'space';
      continue;
    } else if (/[a-z]/.test(char)) {
      charsTypes[i] = 'lowerCase';
    } else if (/[A-Z]/.test(char)) {
      charsTypes[i] = 'upperCase';
    } else if (/[0-9]/.test(char)) {
      charsTypes[i] = 'digit';
    } else {
      charsTypes[i] = 'symbol';
    }

    charsPositions.push(i);
  }

  element.textContent = '';

  let timeout = null;

  (function shuffle(start) {
    const charsArrayCopy = [...charsArray];
    const charsPositionsLength = charsPositions.length;

    if (start > charsPositionsLength) {
      if (typeof options.onComplete === 'function') {
        options.onComplete(element);
      }

      return;
    }

    for (let i = Math.max(start, 0); i < charsPositionsLength; i += 1) {
      if (i < start + options.iterations) {
        charsArrayCopy[charsPositions[i]] = getRandomCharacterForType(charsTypes[charsPositions[i]]);
      } else {
        charsArrayCopy[charsPositions[i]] = '';
      }
    }

    element.textContent = charsArrayCopy.join('');

    timeout = setTimeout(() => {
      shuffle(start + 1);
    }, 1000 / options.fps);
  })(-options.iterations);

  return () => {
    clearTimeout(timeout);
    timeout = null;
  };
}
